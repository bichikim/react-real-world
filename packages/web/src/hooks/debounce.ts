import {debounce, DebounceSettings} from 'lodash'
import {useCallback, useMemo, useRef} from 'react'

const DEFAULT_DELAY = 250

export interface DebounceOptions extends DebounceSettings {
  getter?: (value: any) => any
}

export type DebounceCallback<S> = (value: S) => any

/**
 * debounce calling
 * callback chain
 * @param callback
 * @param wait
 * @param options
 */
export const useDebounce = <S>(
  callback?: DebounceCallback<S>,
  wait: number = DEFAULT_DELAY,
  options: DebounceOptions = {},
) => {
  const _options = useRef(options)
  _options.current = options
  const _callback = useRef(callback)
  _callback.current = callback
  const _stack = useRef<any[]>([])

  const _debounce = useMemo(
    () => debounce((call) => (call()), wait, options),
    [wait, options],
  )

  return useCallback(
    (value) => {
      const {getter} = _options.current
      const result = getter ? getter(value) : value
      _debounce(() => {
        const callback = _callback.current
        callback && callback(result)
        _stack.current = []
      })
    },
    [_debounce],
  )
}
