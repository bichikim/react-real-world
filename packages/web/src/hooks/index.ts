export * from './init'
export * from './fast-memo'
export * from './debounce'
export * from './ref-act'
