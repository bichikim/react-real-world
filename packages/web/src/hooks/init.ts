import {useRef} from 'react'

export const useInit = (initValue) => {
  const init = useRef(initValue)
  return init.current
}
