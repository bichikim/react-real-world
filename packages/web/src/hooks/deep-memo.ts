import isEqual from 'fast-deep-equal'
import {useRef} from 'react'
import {AnyFunction} from 'src/types'

export const useDeepMemo = <T extends AnyFunction>(function_: T, dep): ReturnType<T> => {
  const ref = useRef()
  const result = useRef()

  if (!ref.current) {
    ref.current = {...dep}
    result.current = function_()
    return result.current
  }

  if (isEqual(ref.current, dep)) {
    return result.current
  }

  ref.current = dep
  result.current = function_()
  return result.current
}
