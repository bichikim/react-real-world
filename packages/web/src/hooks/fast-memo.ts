import {FC, memo} from 'react'
import isEqual from 'react-fast-compare'
import {PureObject} from 'src/types'

export const withFastMemo = <P extends PureObject>(Component: FC<P>): FC<P> => {
  return memo(Component, isEqual) as any
}
