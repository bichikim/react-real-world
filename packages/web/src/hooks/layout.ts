import {ComponentType, createElement as h} from 'react'
import {PageFunctionComponent, PureObject} from 'src/types'

export const withLayout = (Layout: ComponentType<PureObject>) =>
  (Page: ComponentType<PureObject>): PageFunctionComponent =>
    (props) => {
      const {children, ...rest} = props
      return (
        h(Layout, {...rest},
          h(Page, {...rest}, children),
        )
      )
    }
