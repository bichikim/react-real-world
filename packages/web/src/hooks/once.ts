import {useMemo} from 'react'
import {AnyFunction} from 'src/types'

export const useOnce = (function_: AnyFunction) => {
  useMemo(() => {
    function_()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
