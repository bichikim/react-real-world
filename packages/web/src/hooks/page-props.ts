import {createContext, useContext, useMemo} from 'react'
import {IS_SSR} from 'src/utils'

export const pagePropsContext = createContext<Record<string, any>>({})

let _pageProps

const useClientPageProps = () => {
  return useMemo(() => {
    const rawJson = document.querySelector('#__NEXT_DATA__')

    if (_pageProps) {
      return _pageProps
    }

    try {
      const result = JSON.parse(rawJson?.innerHTML)
      _pageProps = result?.props?.pageProps ?? {}

      return _pageProps
    } catch {
      return {}
    }
  }, [])
}

export const usePageProps = () => {

  if (IS_SSR) {
    return useContext(pagePropsContext)
  }

  return useClientPageProps()
}
