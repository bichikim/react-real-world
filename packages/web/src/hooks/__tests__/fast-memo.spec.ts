import {act, fireEvent, render} from '@testing-library/react'
import {FC, createElement as h, useState} from 'react'
import {withFastMemo} from 'src/hooks'

const setup = () => {
  const Foo = jest.fn((props) => {
    return h('div', null, props.children)
  })
  const MemoFoo: FC = withFastMemo(Foo)
  const Bar = jest.fn((props) => {
    return h('div', null, props.children)
  })
  const MemoBar: FC = withFastMemo(Bar)

  const Container: FC = () => {
    const [count, setCount] = useState(0)
    return (
      h('div', null,
        h(MemoFoo, null,
          h(MemoBar, null, 'foo'),
        ),
        h('a', {onClick: () => setCount(count + 1)}, 'add'),
        h('div', null, count),
      )
    )
  }

  const wrapper = render(h(Container))

  return {
    ...wrapper,
    Bar,
    Foo,
  }
}

describe('fast-memo', function test() {
  it('should not rerender', function test() {
    const {container, Foo, Bar} = setup()

    expect(Foo.mock.calls.length).toBe(1)
    expect(Bar.mock.calls.length).toBe(1)
    expect(container.firstChild).toMatchSnapshot()

    act(() => {
      const button = container.querySelector('a')
      if (button) {
        fireEvent.click(button)
      }
    })

    expect(Foo.mock.calls.length).toBe(1)
    expect(Bar.mock.calls.length).toBe(1)
    expect(container.firstChild).toMatchSnapshot()
  })
})
