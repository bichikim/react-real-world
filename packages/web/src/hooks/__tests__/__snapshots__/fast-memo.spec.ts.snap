// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fast-memo should not rerender 1`] = `
<div>
  <div>
    <div>
      foo
    </div>
  </div>
  <a>
    add
  </a>
  <div>
    0
  </div>
</div>
`;

exports[`fast-memo should not rerender 2`] = `
<div>
  <div>
    <div>
      foo
    </div>
  </div>
  <a>
    add
  </a>
  <div>
    1
  </div>
</div>
`;
